File: cplusplus/src/create_encrypt.cpp
#include "autograph/create_encrypt.h"

#include "autograph/encrypt.h"

EncryptFunction create_encrypt(const Chunk &our_secret_key) {
  uint32_t index = 0;
  auto encrypt_function = [&our_secret_key, &index](const Chunk &plaintext) {
    index++;
    Chunk ciphertext = encrypt(our_secret_key, index, plaintext);
    return std::move(ciphertext);
  };
  return std::move(encrypt_function);
}

File: typescript/src/create-encrypt.ts
import { concat, fromInteger } from 'stedy/bytes'
import { EncryptFunction } from '../types'
import encrypt from './encrypt'

const createEncrypt = (ourSecretKey: BufferSource): EncryptFunction => {
  let index = 0
  return async (message: BufferSource) => {
    index += 1
    const ciphertext = await encrypt(ourSecretKey, index, message)
    const nonce = fromInteger(index)
    return concat([nonce, ciphertext])
  }
}

export default createEncrypt
